python-package/xgboost/sklearn.py

```diff
@@ -679,3 +679,158 @@ class XGBRegressor(XGBModel, XGBRegressorBase):
+            for the evaluation and value is the value of the evaluation
+            function. This objective is always minimized.
+        early_stopping_rounds : int
+            Activates early stopping. Validation error needs to decrease at
+            least every <early_stopping_rounds> round(s) to continue training.
+            Requires at least one item in evals.  If there's more than one,
+            will use the last. Returns the model from the last iteration
+            (not the best one). If early stopping occurs, the model will
+            have three additional fields: bst.best_score, bst.best_iteration
+            and bst.best_ntree_limit.
+            (Use bst.best_ntree_limit to get the correct value if num_parallel_tree
+            and/or num_class appears in the parameters)
+        verbose : bool
+            If `verbose` and an evaluation set is used, writes the evaluation
+            metric measured on the validation set to stderr.
+        xgb_model : str
+            file name of stored xgb model or 'Booster' instance Xgb model to be
+            loaded before training (allows training continuation).
+        """
+        # check if group information is provided
+        if group is None:
+            raise ValueError("group is required for ranking task")
+
+        if eval_set is not None:
+            if eval_group is None:
+                raise ValueError("eval_group is required if eval_set is not None")
+            elif len(eval_group) != len(eval_set):
+                raise ValueError("length of eval_group should match that of eval_set")
+            elif any(group is None for group in eval_group):
+                raise ValueError("group is required for all eval datasets for ranking tast")
+
+        def _dmat_init(group, **params):
+            ret = DMatrix(**params)
+            ret.set_group(group)
+            return ret
+
+        if sample_weight is not None:
+            train_dmatrix = _dmat_init(group, data=X, label=y, weight=sample_weight,
+                                       missing=self.missing, nthread=self.n_jobs)
+        else:
+            train_dmatrix = _dmat_init(group, data=X, label=y,
+                                       missing=self.missing, nthread=self.n_jobs)
+
+        evals_result = {}
+
+        if eval_set is not None:
+            if sample_weight_eval_set is None:
+                sample_weight_eval_set = [None] * len(eval_set)
+            evals = [_dmat_init(eval_group[i], data=eval_set[i][0], label=eval_set[i][1],
+                                missing=self.missing, weight=sample_weight_eval_set[i],
+                                nthread=self.n_jobs) for i in range(len(eval_set))]
+            nevals = len(evals)
+            eval_names = ["eval_{}".format(i) for i in range(nevals)]
+            evals = list(zip(evals, eval_names))
+        else:
+            evals = ()
+
+        params = self.get_xgb_params()
+
+        if callable(self.objective):
```

Comment:
Are we supporting custom objective for `XGBRanker`? I'm inclined to not supporting it (throw `ValueError` if `self.objective` is callable).

Reply:
@hcho3 Actually, i feel that it might be interesting and beneficial to allow for custom ranking function, but i am not sure how easy it is. Let's disable it for now and I will take a look at it.